File: solutions/haskell/triangle/1/src/Triangle.hs
module Triangle (TriangleType(..), triangleType) where

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Illegal
                  deriving (Eq, Show)

triangleType :: (Num a, Ord a) => a -> a -> a -> TriangleType
triangleType a b c
  | not (isValid a b c) = Illegal
  | isEquilateral a b c = Equilateral
  | isIsosceles a b c = Isosceles
  | otherwise = Scalene
  where
    isValid a b c = (a + b >= c) && (a + c >= b) && (b + c >= a) && (a*b*c > 0)
    isEquilateral a b c = a == b && b == c
    isIsosceles a b c = (a == b) || (a == c) || (b == c)
  

File: solutions/ruby/triangle/1/triangle.rb
class Triangle
  def initialize(arr)
    @a = arr[0]
    @b = arr[1]
    @c = arr[2]
  end
  
  def equilateral?
    triangle? && (@a == @b) && (@a == @c)
  end

  def triangle?
    (@a > 0) & (@b > 0) & (@c > 0) & (@a + @b > @c) & (@b + @c > @a) & (@a + @c > @b)
  end

  def isosceles?
    triangle? & ((@a == @b) | (@a == @c) | (@b == @c))
  end

  def scalene?
    triangle? & (not isosceles?)
  end  
end